network/p2p/cache/node_blocklist_wrapper_test.go

```diff
@@ -0,0 +1,294 @@
+			s.provider.On("ByPeerID", peerID).Return(originalIdentity, true)
+
+			// step 1: before putting node on blocklist,
+			// an Identity with `Ejected` equal to the original value should be returned
+			i, found := s.wrapper.ByNodeID(originalIdentity.NodeID)
+			require.True(s.T(), found)
+			require.Equal(s.T(), originalEjected, i.Ejected)
+
+			i, found = s.wrapper.ByPeerID(peerID)
+			require.True(s.T(), found)
+			require.Equal(s.T(), originalEjected, i.Ejected)
+
+			// step 2: _after_ putting node on blocklist,
+			// an Identity with `Ejected` equal to `true` should be returned
+			err := s.wrapper.Update(flow.IdentifierList{originalIdentity.NodeID})
+			require.NoError(s.T(), err)
+
+			i, found = s.wrapper.ByNodeID(originalIdentity.NodeID)
+			require.True(s.T(), found)
+			require.True(s.T(), i.Ejected)
+
+			i, found = s.wrapper.ByPeerID(peerID)
+			require.True(s.T(), found)
+			require.True(s.T(), i.Ejected)
+
+			// step 3: after removing the node from the blocklist,
+			// an Identity with `Ejected` equal to the original value should be returned
+			err = s.wrapper.Update(flow.IdentifierList{})
+			require.NoError(s.T(), err)
+
+			i, found = s.wrapper.ByNodeID(originalIdentity.NodeID)
+			require.True(s.T(), found)
+			require.Equal(s.T(), originalEjected, i.Ejected)
+
+			i, found = s.wrapper.ByPeerID(peerID)
+			require.True(s.T(), found)
+			require.Equal(s.T(), originalEjected, i.Ejected)
+		})
+	}
+}
+
+// TestUpdate tests updating, clearing and retrieving the blocklist
+// Note: conceptually, the blocklist is a set, i.e. not order dependent.
+// The wrapper internally converts the list to a set and vice versa. Therefore
+// the order is not preserved by `GetBlocklist`. Consequently, we compare
+// map-based representations here.
+func (s *NodeBlocklistWrapperTestSuite) TestUpdate() {
+	blocklist1 := unittest.IdentifierListFixture(8)
+	blocklist2 := unittest.IdentifierListFixture(11)
+	blocklist3 := unittest.IdentifierListFixture(5)
+
+	err := s.wrapper.Update(blocklist1)
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), blocklist1.Lookup(), s.wrapper.GetBlocklist().Lookup())
+
+	err = s.wrapper.Update(blocklist2)
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), blocklist2.Lookup(), s.wrapper.GetBlocklist().Lookup())
+
+	err = s.wrapper.ClearBlocklist()
+	require.NoError(s.T(), err)
+	require.Empty(s.T(), s.wrapper.GetBlocklist())
+
+	err = s.wrapper.Update(blocklist3)
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), blocklist3.Lookup(), s.wrapper.GetBlocklist().Lookup())
+}
+
+// TestDataBasePersist verifies that blocklist is persisted in database
+func (s *NodeBlocklistWrapperTestSuite) TestDataBasePersist() {
+	blocklist := unittest.IdentifierListFixture(8)
+
+	// update blocklist and check DB that the new value is there
+	err := s.wrapper.Update(blocklist)
+	require.NoError(s.T(), err)
+
+	var b1 map[flow.Identifier]struct{}
+	err = s.DB.View(operation.RetrieveBlocklist(&b1))
```

Comment:
This would imply that we know the implementation detail of database. And therefore coupled by the database design, if we change design, this test would have to change as well. 

The database implementation is better to be tested by their own unit test. 

So here, a better strategy is to create a new block list wrapper that wraps the database. Since a new wrapper is created, it should have forgotten what is possibly being cached in memory and if it returns the correct data, then it must be from the database.

Reply:
Thanks for the great suggestion on how to decouple testing of `NodeBlocklistWrapper` from the lower-level data base design!
* Added separate [unity tests](https://github.com/onflow/flow-go/blob/426c830174ecdb34b277fea56d9112a67cca4e9c/storage/badger/operation/bft_test.go#L18) for the lower-level data-base abstraction
* Refactored the unit tests for `NodeBlocklistWrapper` to implement you suggestion
   > create a new block list wrapper that wraps the database. Since a new wrapper is created, it should have forgotten what is possibly being cached in memory and if it returns the correct data, then it must be from the database.

   updated test: [`TestDataBasePersist`](https://github.com/onflow/flow-go/blob/4b252e1e4377db871dc6b3e312e73fa9c5e92402/network/p2p/cache/node_blocklist_wrapper_test.go#L265-L276)